core/util/fixed_merkle_tree.go

```diff
@@ -29,78 +82,140 @@ func NewFixedMerkleTree(chunkSize int) *FixedMerkleTree {
-	for n := 0; n < 1024; n++ {
-		fmt.Leaves[n] = NewCompactMerkleTree(nil)
+	fmt.Leaves = make([]Hashable, FixedMerkleLeaves)
+	for i := 0; i < FixedMerkleLeaves; i++ {
+		fmt.Leaves[i] = getNewLeaf()
 	}
 }
 
-func (fmt *FixedMerkleTree) Write(buf []byte, chunkIndex int) error {
-	//split chunk into 1024 parts for challenge hash
-	merkleChunkSize := fmt.ChunkSize / 1024
+// writeToLeaves will divide the data with MerkleChunkSize(64 bytes) and write to
+// each leaf hasher
+func (fmt *FixedMerkleTree) writeToLeaves(b []byte) error {
+	if len(b) > MaxMerkleLeavesSize {
+		return goError.New("data size greater than maximum required size")
+	}
 
-	// chunksize is less than 1024
-	if merkleChunkSize == 0 {
-		merkleChunkSize = 1
+	if len(b) < MaxMerkleLeavesSize && !fmt.isFinal {
+		return goError.New("invalid merkle leaf write")
 	}
 
-	total := len(buf)
-	offset := 0
-	for i := 0; i < total; i += merkleChunkSize {
-		end := i + merkleChunkSize
-		if end > len(buf) {
-			end = len(buf)
+	leafInd := 0
+	for i := 0; i < len(b); i += MerkleChunkSize {
+		j := i + MerkleChunkSize
+		if j > len(b) {
+			j = len(b)
 		}
 
-		if len(fmt.Leaves) != 1024 {
-			fmt.initLeaves()
-		}
+		fmt.Leaves[leafInd].Write(b[i:j])
+		leafInd++
+	}
 
-		err := fmt.Leaves[offset].AddDataBlocks(buf[i:end], chunkIndex)
-		if errors.Is(err, ErrLeafNoSequenced) {
-			return err
-		}
+	return nil
+}
+
+// Write will write data to the leaves once MaxMerkleLeavesSize(64 KB) is reached.
+// This can be used to write stream of data as well.
+// fmt.Finalize() is required if data is not exact multiple of 64KB.
+func (fmt *FixedMerkleTree) Write(b []byte) (int, error) {
+
+	fmt.writeLock.Lock()
+	defer fmt.writeLock.Unlock()
+	if fmt.isFinal {
+		return 0, goError.New("cannot write. Tree is already finalized")
+	}
 
-		offset++
-		if offset >= 1024 {
-			offset = 0
+	for i, j := 0, MaxMerkleLeavesSize-fmt.writeCount; i < len(b); i, j = j, j+MaxMerkleLeavesSize {
+		if j > len(b) {
+			j = len(b)
+		}
+		prevWriteCount := fmt.writeCount
+		fmt.writeCount += int(j - i)
+		copy(fmt.writeBytes[prevWriteCount:fmt.writeCount], b[i:j])
+		if fmt.writeCount == MaxMerkleLeavesSize {
+			err := fmt.writeToLeaves(fmt.writeBytes)
+			if err != nil {
+				return 0, err
+			}
+			fmt.writeCount = 0
 		}
 	}
+	return len(b), nil
+}
 
+// GetMerkleRoot is only for interface compliance.
+func (fmt *FixedMerkleTree) GetMerkleTree() MerkleTreeI {
 	return nil
 }
 
-// GetMerkleRoot get merkle tree
-func (fmt *FixedMerkleTree) GetMerkleTree() MerkleTreeI {
-	merkleLeaves := make([]Hashable, 1024)
+func (fmt *FixedMerkleTree) CalculateMerkleRoot() {
+	nodes := make([][]byte, len(fmt.Leaves))
+	for i := 0; i < len(nodes); i++ {
+		nodes[i] = fmt.Leaves[i].GetHashBytes()
+	}
+
+	for i := 0; i < FixedMTDepth; i++ {
 
-	for idx, leaf := range fmt.Leaves {
+		newNodes := make([][]byte, (len(nodes)+1)/2)
+		if len(nodes)&1 == 1 {
+			nodes = append(nodes, nodes[len(nodes)-1])
```

Comment:
Like you said, I think of using hash as it is to be more appropriate.
It was the existing logic to duplicate the last node. 
I could have removed the duplication logic because the fixedMerkleTree will have exactly 1024 leaves which is 2^10 i.e. no odd number of nodes in any level.

Will update it.